### ISSUE TITLE
Degraded Performance Converting Transformer Engine-Trained LM to HuggingFace

### ISSUE THREAD

**Issue by Kyle1668** [EleutherAI/gpt-neox#1353]
**Describe the bug**
I recently trained a 6.9B LM where I leveraged some of the new Transformer Engine features. These include TE implementations of MLP blocks with layer normalization and the TE implementation of MHA (`te_layernorm_mlp` and `te_mha`). I wish to convert this model to HuggingFace format. I have run into two challenges. 

### 1. Lack of Layernorm in MLP in `GPTNeoXForCausalLM`

My understanding is that the HF implementation of NeoX does not have layer normalization within the MLP modules. This is blocking conversion from working out of the box. I have developed a workaround by creating a new custom type:

```python
from torch import nn
from transformers import GPTNeoXForCausalLM, AutoModelForCausalLM, GPTNeoXConfig
from transformers.models.gpt_neox.modeling_gpt_neox import GPTNeoXMLP


class TransformerEngineGPTNeoXMLP(GPTNeoXMLP):
    """Custom MLP with LayerNorm for Transformer Engine compatibility"""

    def __init__(self, config):
        super().__init__(config)
        # Add the LayerNorm that Transformer Engine uses
        self.layer_norm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps, elementwise_affine=True)

    def forward(self, hidden_states):
        # Apply layer norm before the dense_h_to_4h (FC1)
        hidden_states = self.layer_norm(hidden_states)
        # Then proceed with the original MLP forward pass
        hidden_states = self.dense_h_to_4h(hidden_states)
        hidden_states = self.act(hidden_states)
        hidden_states = self.dense_4h_to_h(hidden_states)
        return hidden_states


class TransformerEngineGPTNeoXForCausalLM(GPTNeoXForCausalLM):
    """GPTNeoX with custom MLP that includes LayerNorm for Transformer Engine compatibility"""

    def __init__(self, config):
        super().__init__(config)
        # Replace all MLP instances with our custom MLP
        for i, layer in enumerate(self.gpt_neox.layers):
            self.gpt_neox.layers[i].mlp = TransformerEngineGPTNeoXMLP(config)

    @classmethod
    def from_config(cls, config):
        """Creates a TransformerEngineGPTNeoXForCausalLM from a configuration."""
        return cls(config)


# Register our custom model with the AutoModel classes
AutoModelForCausalLM.register(
    GPTNeoXConfig, TransformerEngineGPTNeoXForCausalLM, exist_ok=True
)
```

```python
import torch
from custom_neox_models import TransformerEngineGPTNeoXForCausalLM

model_path = "/data/checkpoints/hf_converted/annealing_baseline/global_step11920/"
model = TransformerEngineGPTNeoXForCausalLM.from_pretrained(
    model_path,
    torch_dtype=torch.bfloat16,
).to("cuda:0")
model
```

```
Loading checkpoint shards:  33%|███▎      | 2/6 [00:01<00:02,  1.36it/s]
Loading checkpoint shards: 100%|██████████| 6/6 [00:03<00:00,  1.50it/s]
TransformerEngineGPTNeoXForCausalLM(
  (gpt_neox): GPTNeoXModel(
    (embed_in): Embedding(50304, 4096)
    (emb_dropout): Dropout(p=0.0, inplace=False)
    (layers): ModuleList(
      (0-31): 32 x GPTNeoXLayer(
        (input_layernorm): LayerNorm((4096,), eps=1e-05, elementwise_affine=True)
        (post_attention_layernorm): LayerNorm((4096,), eps=1e-05, elementwise_affine=True)
        (post_attention_dropout): Dropout(p=0.0, inplace=False)
        (post_mlp_dropout): Dropout(p=0.0, inplace=False)
        (attention): GPTNeoXAttention(
          (query_key_value): Linear(in_features=4096, out_features=12288, bias=True)
          (dense): Linear(in_features=4096, out_features=4096, bias=True)
        )
        (mlp): TransformerEngineGPTNeoXMLP(
          (dense_h_to_4h): Linear(in_features=4096, out_features=16384, bias=True)
          (dense_4h_to_h): Linear(in_features=16384, out_features=4096, bias=True)
          (act): GELUActivation()
          (layer_norm): LayerNorm((4096,), eps=1e-05, elementwise_affine=True)
        )
      )
    )
    (final_layer_norm): LayerNorm((4096,), eps=1e-05, elementwise_affine=True)
    (rotary_emb): GPTNeoXRotaryEmbedding()
  )
  (embed_out): Linear(in_features=4096, out_features=50304, bias=False)
)
```

### 2. Degraded Performance

Using the custom model class from the previous step is able to get the model to somewhat function proerly. However, I find that the model descends into gibberish and repetitions after generating ~40 tokens.

```python
inputs = tokenizer("Spiderman is a", return_tensors="pt", return_token_type_ids=False).to("cuda:0")
outputs = model.generate(**inputs, max_new_tokens=100, do_sample=False)
print(tokenizer.batch_decode(outputs, skip_special_tokens=False)[0])
```

```
Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.
Spiderman is a fictional character, a superhero that appears in comic books published by Marvel Comics. He is a human who gained spider-like abilities after being bitten by a radioactive spider. He is a high school student who is a crime-fighting vigilante who wears a red and blue costume and mask. He is a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a
``` 

Benchmark performance is also regressed. For example, the model gets ~53% accuracy on WMDP Bio using the NeoX `eval.py` interface for the EleutherAI LM Eval Harness, but converted model only gets ~38%. 

```python
import lm_eval
from lm_eval.models.huggingface import HFLM

lm_eval_model = HFLM(model)
task_manager = lm_eval.tasks.TaskManager()
qa_benchmark_results = lm_eval.simple_evaluate(
    model=lm_eval_model,
    tasks=["wmdp_bio"],
    num_fewshot=0,
    task_manager=task_manager,
    batch_size=128,
)
qa_benchmark_results
```

```
`pretrained` model kwarg is not of type `str`. Many other model arguments may be ignored. Please do not launch via accelerate or use `parallelize=True` if passing an existing model this way.
Passed an already-initialized model through `pretrained`, assuming single-process call to evaluate() or custom distributed integration
[Task: wmdp_bio] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
[Task: wmdp_bio] has_training_docs and has_validation_docs are False, using test_docs as fewshot_docs but this is not recommended.
100%|██████████| 1273/1273 [00:01<00:00, 992.95it/s] 
Running loglikelihood requests: 100%|██████████| 5092/5092 [00:16<00:00, 313.43it/s]

{'results': {'wmdp_bio': {'alias': 'wmdp_bio',
   'acc,none': 0.3857030636292223,
   'acc_stderr,none': 0.013648098974225583}},
 'group_subtasks': {'wmdp_bio': []},
 'configs': {'wmdp_bio': {'task': 'wmdp_bio',
   'dataset_path': 'cais/wmdp',
   'dataset_name': 'wmdp-bio',
   'test_split': 'test',
   'doc_to_text': '{{question.strip()}}\nA. {{choices[0]}}\nB. {{choices[1]}}\nC. {{choices[2]}}\nD. {{choices[3]}}\nAnswer:',
   'doc_to_target': 'answer',
   'unsafe_code': False,
   'doc_to_choice': ['A', 'B', 'C', 'D'],
   'description': 'The following are multiple choice questions (with answers) about biology.\n\n',
   'target_delimiter': ' ',
   'fewshot_delimiter': '\n\n',
   'num_fewshot': 0,
   'metric_list': [{'metric': 'acc',
     'aggregation': 'mean',
     'higher_is_better': True}],
   'output_type': 'multiple_choice',
   'repeats': 1,
   'should_decontaminate': False,
   'metadata': {'version': 1}}},
 'versions': {'wmdp_bio': 1},
 'n-shot': {'wmdp_bio': 0},
...
 'tokenizer_pad_token': ['<|pad|>', '50277'],
 'tokenizer_eos_token': [None, 'None'],
 'tokenizer_bos_token': [None, 'None'],
 'eot_token_id': None,
 'max_length': 2048}
```

**To Reproduce**

This script handles the conversion. It is a modification of the existing conversion script: https://github.com/EleutherAI/gpt-neox/blob/filtering_training/tools/ckpts/convert_transformer_engine_neox_to_hf.py

This notebooks tests the converted checkpoint: https://github.com/EleutherAI/gpt-neox/blob/filtering_training/tools/ckpts/converted_checkpoint_test.ipynb

**Expected behavior**
I expect the model to be able to generate far more tokens while remaining coherent. I expect benchmark performance post-conversion to be consistent with the original NeoX checkpoint. 

**Proposed solution**

I think my solution largely sidesteps the performance regressions from the incomplete MLP module. The other primary transformer engine feature enabled is `te_mha`, perhaps there is a difference in the implementations?

I think weights not being converted correctly is an unlikely issue. I've run a number of tests which suggest that the weights in the HF model are identical to those in the model states files. There could be a bug in my update to the conversion script, however. 

**Environment (please complete the following information):**
 - GPUs: 8xH100
- Configs:

```yaml
{
  # Tokens
  "data_path": "/data/pretraining-mix/pretraining-mix_text_document",
  "vocab_file": "/data/neox_tokenizer/tokenizer.json",
  "tokenizer_type": "HFTokenizer",
  "data_impl": "mmap",

  # Logging
  "checkpoint_validation_with_forward_pass": False,
  "tensorboard_dir": "tensorboard",
  "log_dir": "logs",
  "log_interval": 10,
  "steps_per_print": 10,
  "wall_clock_breakdown": true,
  "use_wandb": True,
  "wandb_host": "https://api.wandb.ai",
  "wandb_project": "AISI",
  "wandb_team": "eleutherai",
  "wandb_run_name": "pretraining_baseline",

  # Distributed Training
  "hostfile": "/workspace/hostfile",
  "deepspeed_mpi": True,
  "launcher": "openmpi",
  "deepspeed_extra_args": { "ssh_port": 2222 },
  "pipe_parallel_size": 1,
  "model_parallel_size": 1,

  # Training Duration
  # 500B (tokens) / (4 (grad acc) * 32 (world size) * 16 (micro batch size) * 2048 (seq length))
  "train_iters": 119209,
  "lr_decay_iters": 119209,
  "distributed_backend": "nccl",
  "lr_decay_style": "cosine",
  "warmup": 0.01,
  "split": "100,0,0",

  # Architecture
  "num_layers": 32,
  "hidden_size": 4096,
  "num_attention_heads": 32,
  "seq_length": 2048,
  "max_position_embeddings": 2048,
  "norm": "layernorm",
  "pos_emb": "rotary",
  "rotary_pct": 0.25,
  "no_weight_tying": true,
  "gpt_j_residual": true,
  "output_layer_parallelism": "column",
  "attention_config": [[["flash"], 32]],
  "scaled_upper_triang_masked_softmax_fusion": true,
  "precision": "bfloat16",
  "activation": "gelu",
  # "activation": "swiglu", TODO: Support SwiGLU

  # Transformer Engine
  "te_columnparallel": false,
  "te_rowparallel": false,
  "te_layernorm_mlp": true,
  "te_mha": true,
  "te_fp8_format": "hybrid",
  "te_fp8_wgrad": false,
  "te_fp8_amax_history_len": 1,
  "te_fp8_amax_compute_algo": "most_recent",
  "te_fp8_margin": 0,
  "te_fp8_mha": false,

  # Optimization
  # 0.0003 is OLMo 2's peak learning rate
  "optimizer":
    {
      "type": "Adam",
      "params": { "lr": 0.0003, "betas": [0.9, 0.95], "eps": 1.0e-8 },
    },
  "min_lr": 0.000012,
  "zero_optimization":
    {
      "stage": 1,
      "allgather_partitions": true,
      "allgather_bucket_size": 1260000000,
      "overlap_comm": true,
      "reduce_scatter": true,
      "reduce_bucket_size": 1260000000,
      "contiguous_gradients": true,
      "cpu_offload": false,
    },
  "train_micro_batch_size_per_gpu": 16,
  "gradient_accumulation_steps": 4,
  "gradient_clipping": 1.0,
  "weight_decay": 0.1,
  "hidden_dropout": 0,
  "attention_dropout": 0,

  # Checkpointing
  "checkpoint_activations": true,
  "checkpoint_num_layers": 1,
  "partition_activations": true,
  "synchronize_each_layer": true,
  "checkpoint_factor": 2000,
  "save": "/checkpoints/pretraining_baseline",
  "load": "/checkpoints/pretraining_baseline",

  # Evaluation
  "eval_iters": 0,
  "eval_interval": 5,
  "eval_results_prefix": "pretraining_baseline",
  # TODO: Add automatic evals for next runs
  #"eval_tasks": ["wmdp_bio"],
  # "do_valid",
  # "eval_tasks": ["wmdp", "mmlu", "piqa", "arc-easy", "arc-challenge", "lambada", "hellaswag"]
}

```

**Comment by aurelion-source** [comment]
Hi @Kyle1668!

GPTNeoXForCausalLM does have layer normalization within the MLP module.
To clarify, `post_attention_layernorm.weight` and `layer_norm_weight` should refer to the same named parameter.

So in your solution, it appears that the issue is coming from `post_attention_layernorm` being applied twice.
1. Once inside the HuggingFace model (with randomly initialized weights that aren’t used during training):
https://github.com/huggingface/transformers/blob/6daa3eeba582facb57cd71db8efb66998b12942f/src/transformers/models/gpt_neox/modular_gpt_neox.py#L257
2. And once again in your custom class here: `hidden_states = self.layer_norm(hidden_states)`

To fix this cleanly, I updated the conversion script to directly assign the trained TELayerNorm weights to post_attention_layernorm in the HF model. That way, you can use the standard GPTNeoXForCausalLM without needing the custom class, and avoid double-layernorm issues.

Additionally, to avoid future confusion, post_attention_layernorm is now only initialized if te_lnmlp is not used.
That said, models already trained with the layer should still convert fine as explained here: #1355

Let me know if you run into anything else!
